ngraph/core/reference/include/ngraph/runtime/reference/convolution.hpp

```diff
@@ -24,338 +25,347 @@
-            struct widen<float>
-            {
-                using type = double;
-            };
+                    Tensor& operator++()
+                    {
+                        buffer += size;
+                        return *this;
+                    }
+                };
 
-            template <>
-            struct widen<double>
-            {
-                using type = long double;
-            };
+                template <typename Buf>
+                Tensor<Buf> tensor(Buf buffer, const Shape& s)
+                {
+                    static_assert(std::is_pointer<Buf>::value, "expecting ptr");
+                    return Tensor<Buf>{buffer, s, shape_size(s)};
+                }
 
-            // in: NC_I...
-            // filter: C_OC_I...
-            // out: NC_O...
-            template <typename INPUT,
-                      typename FILTER,
-                      typename OUTPUT,
-                      typename ACCUMULATION = typename widen<OUTPUT>::type>
-            void general_convolution(const INPUT* in,
-                                     const FILTER* filter,
-                                     OUTPUT* out,
-                                     const Shape& in_shape,
-                                     const Shape& filter_shape,
-                                     const Shape& out_shape,
-                                     const Strides& stride,
-                                     const Strides& filter_dilation,
-                                     const CoordinateDiff& in_pad_below,
-                                     const CoordinateDiff& in_pad_above,
-                                     const Strides& in_dilation,
-                                     size_t in_batch_axis,
-                                     size_t in_channel_axis,
-                                     size_t filter_out_channel_axis,
-                                     size_t filter_in_channel_axis,
-                                     size_t out_batch_axis,
-                                     size_t out_channel_axis)
-            {
-                auto old_mode = std::fegetround();
-                std::fesetround(FE_TONEAREST);
-                // Comments throughout assume without loss of generality that:
-                //
-                // * batch axes for both in and out are 0
-                // * in channel axes for both in and filter are 1
-                // * out channel axes for filter is 0
-                // * out channel axis for out is 1
+                template <typename Tens>
+                Tens tensor_slice(Tens t)
+                {
+                    Shape new_shape(++t.shape.begin(), t.shape.end());
+                    return tensor(t.buffer, new_shape);
+                }
 
-                // At the outermost level we will walk over every out coordinate O.
-                CoordinateTransform out_transform(out_shape);
+                enum class ConvolutionType
```

Comment:
Looks like we  don't need `ConvolutionType`  enumeration because we can understand it from the filter rank.

Reply:
It's only for readability. Can be removed if you want.

Reply:
As we agreed in offline talk I will remove it.

Reply:
Removed in this [commit](https://github.com/openvinotoolkit/openvino/pull/3922/commits/cc79afea07bab9f9342f846e7d0eee4728ecf735).